src/model/openthaigpt_pretraining_model/models/llama_hf/model.py

```diff
@@ -0,0 +1,508 @@
+from transformers import (
+    LlamaConfig,
+    LlamaForCausalLM,
+)
+from transformers.models.llama.modeling_llama import (
+    LlamaPreTrainedModel,
+    LlamaDecoderLayer,
+    LlamaRMSNorm,
+    _make_causal_mask,
+    _expand_mask,
+    logger,
+)
+
+import torch
+from torch import nn
+from torch.nn import CrossEntropyLoss
+
+from typing import Optional, Tuple, List, Union
+from transformers.modeling_outputs import (
+    BaseModelOutputWithPast,
+    CausalLMOutputWithPast,
+)
+
+
+class LlamaModel(LlamaPreTrainedModel):
```

Comment:
How different between `LlamaModel` in this code and from `huggingface trasformers`? If you modified some function, please use class inherit.

Reply:
I modified init to use new self.layers -> LlamaDecoderLayerWithCheckpointing(config) and forward to pass gradient_checkpointing parameters.

Reply:
done